### DISCUSSION TITLE
Get play count from last.fm and update the DB

### DISCUSSION THREAD

**Discussion by CustardFlan** [navidrome/navidrome#3454] [Ideas]
Hi!
First, navidrome is awesome! thanks!

Second, i have a huge music library (~58000 tracks) i started replacing it only with FLAC versions
Everytime i replace an album, the playcount, favorites and rating is lost

I been keeping track of my listening habits with last.fm since 2012, for me, it's very valuable data, but, every music software i tried fails to integrate that data with my library, i'm solely interested in the play counts

Investigating a little bit, there is an API endpoint from last.fm that can get these data (including the "loved" flag)

For example:
https://ws.audioscrobbler.com/2.0/?autocorrect=1&username=CustardSandS&format=json&api_key=THEAPIKEY&method=track.getInfo&artist=Daft+Punk&track=WDPK+83.7+FM

That returns something like this:
"userplaycount":"2","userloved":"0"

Maybe, with this data, the playcount and favs can be synced or updated to the navidrome DB, in the user account that have enabled the option, or updated by request, or on every track addition and track loading, for keep the DB in sync with last.fm

I'm sure that it's an nice addition to the general functionality of the software!

**Comment by feamsr00** [comment]
I had just come to request the same. I'm coming from MusicBee and have considered replacing it with this. Thing is [MB has been syncing](https://musicbee.fandom.com/wiki/Tagging_Tools#Last.fm) all my library stats for ~17k tracks over 10+yrs to last.fm. So, all the "discovery" features of Subsonic clients really need that data.

**Comment by kalon33** [comment]
Could also be nice to support Listenbrainz for that, to be able to cover all use cases, and to choose between them.

**Comment by MRodneyMcKay** [comment]
It is a splendid idea, i would love it!
Being able to sync this data after a catastrophic failure would be good. As a bonus, this would be useful, if you have listened to seome music on spotify and once you import the music to navidrome, it could have your plays statistcs ready for it

**Comment by pyxfluff** [comment]
hoping this is still being looked at, i just lost my database again and need to rework everything myself :cry:

**Comment by breakfast-first** [comment]
I would also like this feature. 

I found a workaround using MusicBee as intermediary. Get MusicBee plugin [Sync with LastFm Playcount](https://www.getmusicbee.com/addons/plugins/491/sync-with-lastfm-playcount/) to sync your LastFm play counts to your MB library, then follow the instructions for [MusicBee to Navidrome Sync (MBNDS)](https://github.com/rombat/musicbee-navidrome-sync?tab=readme-ov-file) to sync MB play counts to your ND database. By default it replaces the ND play count only if it's higher, but there's also an option to have it add the play count values instead of replacing them. I successfully updated my play counts for 1 artist as a test and am running it on my entire library now.

**Comment by zeroquinc** [comment]
I created a simple python script with the help of AI, it checks all your last.fm scrobbles, fetches the play count, and compares it to the Navidrome database if it finds a song that matches. Then updates the play count in Navidrome. It also updates the last played date on the track and loves/likes the track on Navidrome if its loved on last.fm.

It will also save 2 json files, missing_from_loved.json and missing_from_scrobbles.json. These will include artist, album and track names that you are missing in your Navidrome library compared to your last.fm history.

Worked pretty good for me, but I have only 50k scrobbles and 4k tracks. You can test it for yourself!
It's pretty awesome to have my last.fm and navidrome play count in sync now.

Keep in mind that last.fm plays are leading, it will replace the navidrome plays with the total plays in last.fm.
You can find your User ID in the url when you go to your userpage on the Navidrome page.
It looks like this:  http://localhost:4533/app/#/user/3MQeo7X8DqcVg3hY5V742a

Also 100% make sure Navidrome isn't running when you run this script, and make a backup first.

⚠️ **ALWAYS MAKE A BACKUP OF YOUR NAVIDROME.DB FIRST** ⚠️

I made a repo for it:

https://github.com/zeroquinc/NaviSync

⚠️ **ALWAYS MAKE A BACKUP OF YOUR NAVIDROME.DB FIRST** ⚠️

```
Updated Patrick Doyle - Neville's Waltz to 16 plays, last played 2025-10-15 18:22:17+00:00
Updated Patrick Doyle - Harry in Winter to 17 plays, last played 2025-10-16 12:37:11+00:00
Updated Patrick Doyle - Rita Skeeter to 5 plays, last played 2025-10-05 17:44:10+00:00
Updated Patrick Doyle - Voldemort to 1 plays, last played 2025-10-05 14:07:01+00:00
Updated Patrick Doyle - Underwater Secrets to 4 plays, last played 2025-10-13 13:30:35+00:00
Updated Patrick Doyle - Harry Sees Dragons to 4 plays, last played 2025-10-05 17:47:56+00:00
Updated Patrick Doyle - The Dark Mark to 5 plays, last played 2025-10-05 17:37:49+00:00
Updated Patrick Doyle - The Quidditch World Cup to 98 plays, last played 2025-10-15 23:25:42+00:00
Updated Patrick Doyle - The Maze to 1 plays, last played 2025-09-25 21:39:17+00:00
Updated Patrick Doyle - Another Year Ends to 14 plays, last played 2025-10-15 14:51:17+00:00
Updated Patrick Doyle - Hogwarts' March to 4 plays, last played 2025-10-15 18:27:54+00:00
Updated Patrick Doyle - Sirius Fire to 3 plays, last played 2025-10-05 17:46:06+00:00
Updated Patrick Doyle - Hogwarts' Hymn to 53 plays, last played 2025-10-14 16:07:13+00:00
Updated Patrick Doyle - Death of Cedric to 2 plays, last played 2025-10-16 00:37:56+00:00
Updated Patrick Doyle - Frank Dies to 1 plays, last played 2025-10-05 14:27:12+00:00
Updated Patrick Doyle - Foreign Visitors Arrive to 41 plays, last played 2025-10-17 21:39:06+00:00
Updated Patrick Doyle - The Goblet of Fire to 3 plays, last played 2025-10-05 14:12:24+00:00
Updated Patrick Doyle - The Black Lake to 1 plays, last played 2025-10-05 14:20:18+00:00
Updated Patrick Doyle - The Story Continues to 3 plays, last played 2025-10-05 14:44:35+00:00
Updated Patrick Doyle - Potter Waltz to 18 plays, last played 2025-10-13 23:01:29+00:00
```

**Reply by KruyKaze** [reply]
> I created a simple python script with the help of AI, it checks all your last.fm scrobbles, fetches the play count, and compares it to the Navidrome database if it finds a song that matches. Then updates the play count in Navidrome. It also updates the last played date on the track.
> 
> Worked pretty good for me, but I have only 50k scrobbles and 4k tracks. You can test it for yourself!
> 
> ⚠️ **ALWAYS MAKE A BACKUP OF YOUR NAVIDROME.DB FIRST** ⚠️
> 
> https://gist.github.com/zeroquinc/2809859f64a8146ed18f6d7fa64e7cf1
> 
> ⚠️ **ALWAYS MAKE A BACKUP OF YOUR NAVIDROME.DB FIRST** ⚠️
> 
> ```
> Updated Patrick Doyle - Neville's Waltz to 16 plays, last played 2025-10-15 18:22:17+00:00
> Updated Patrick Doyle - Harry in Winter to 17 plays, last played 2025-10-16 12:37:11+00:00
> Updated Patrick Doyle - Rita Skeeter to 5 plays, last played 2025-10-05 17:44:10+00:00
> Updated Patrick Doyle - Voldemort to 1 plays, last played 2025-10-05 14:07:01+00:00
> Updated Patrick Doyle - Underwater Secrets to 4 plays, last played 2025-10-13 13:30:35+00:00
> Updated Patrick Doyle - Harry Sees Dragons to 4 plays, last played 2025-10-05 17:47:56+00:00
> Updated Patrick Doyle - The Dark Mark to 5 plays, last played 2025-10-05 17:37:49+00:00
> Updated Patrick Doyle - The Quidditch World Cup to 98 plays, last played 2025-10-15 23:25:42+00:00
> Updated Patrick Doyle - The Maze to 1 plays, last played 2025-09-25 21:39:17+00:00
> Updated Patrick Doyle - Another Year Ends to 14 plays, last played 2025-10-15 14:51:17+00:00
> Updated Patrick Doyle - Hogwarts' March to 4 plays, last played 2025-10-15 18:27:54+00:00
> Updated Patrick Doyle - Sirius Fire to 3 plays, last played 2025-10-05 17:46:06+00:00
> Updated Patrick Doyle - Hogwarts' Hymn to 53 plays, last played 2025-10-14 16:07:13+00:00
> Updated Patrick Doyle - Death of Cedric to 2 plays, last played 2025-10-16 00:37:56+00:00
> Updated Patrick Doyle - Frank Dies to 1 plays, last played 2025-10-05 14:27:12+00:00
> Updated Patrick Doyle - Foreign Visitors Arrive to 41 plays, last played 2025-10-17 21:39:06+00:00
> Updated Patrick Doyle - The Goblet of Fire to 3 plays, last played 2025-10-05 14:12:24+00:00
> Updated Patrick Doyle - The Black Lake to 1 plays, last played 2025-10-05 14:20:18+00:00
> Updated Patrick Doyle - The Story Continues to 3 plays, last played 2025-10-05 14:44:35+00:00
> Updated Patrick Doyle - Potter Waltz to 18 plays, last played 2025-10-13 23:01:29+00:00
> ```

does it sync the favorite status of tracks by chance? I too lost all my library and working on getting the metadata back using lastfm.
Thanks

**Reply by zeroquinc** [reply]
> 
> does it sync the favorite status of tracks by chance? I too lost all my library and working on getting the metadata back using lastfm. Thanks

Initially it didn't, because I never love tracks in last.fm, but I saw that the API endpoint was pretty similar so I added it in.
I tested it and it worked fine!

I edited the gist.

**Reply by zeroquinc** [reply]
I updated it another time because I noticed all of my tracks with featurings on it didnt sync, because I always scrobble with only the first artist. So I added a option for that and a whitelist.

**Reply by KruyKaze** [reply]
> > does it sync the favorite status of tracks by chance? I too lost all my library and working on getting the metadata back using lastfm. Thanks
> 
> Initially it didn't, because I never love tracks in last.fm, but I saw that the API endpoint was pretty similar so I added it in. I tested it and it worked fine!
> 
> I edited the gist.

Excellent! what happens when the track is in lastfm but not yet present in my library since i'm rebuilding it?

**Reply by zeroquinc** [reply]
> > > does it sync the favorite status of tracks by chance? I too lost all my library and working on getting the metadata back using lastfm. Thanks
> > 
> > 
> > Initially it didn't, because I never love tracks in last.fm, but I saw that the API endpoint was pretty similar so I added it in. I tested it and it worked fine!
> > I edited the gist.
> 
> Excellent! what happens when the track is in lastfm but not yet present in my library since i'm rebuilding it?

Then it skips it. It will only match with tracks in your current library. You can just re-run it later when you have more albums imported.

**Reply by KruyKaze** [reply]
> > > > does it sync the favorite status of tracks by chance? I too lost all my library and working on getting the metadata back using lastfm. Thanks
> > > 
> > > 
> > > Initially it didn't, because I never love tracks in last.fm, but I saw that the API endpoint was pretty similar so I added it in. I tested it and it worked fine!
> > > I edited the gist.
> > 
> > 
> > Excellent! what happens when the track is in lastfm but not yet present in my library since i'm rebuilding it?
> 
> Then it skips it. It will only match with tracks in your current library. You can just re-run it later when you have more albums imported.

Let me give it a try . Should I leave this alone? "SCROBBLES_JSON = "scrobbles.json" "

**Reply by zeroquinc** [reply]
> > > > > does it sync the favorite status of tracks by chance? I too lost all my library and working on getting the metadata back using lastfm. Thanks
> > > > 
> > > > 
> > > > Initially it didn't, because I never love tracks in last.fm, but I saw that the API endpoint was pretty similar so I added it in. I tested it and it worked fine!
> > > > I edited the gist.
> > > 
> > > 
> > > Excellent! what happens when the track is in lastfm but not yet present in my library since i'm rebuilding it?
> > 
> > 
> > Then it skips it. It will only match with tracks in your current library. You can just re-run it later when you have more albums imported.
> 
> Let me give it a try . Should I leave this alone? "SCROBBLES_JSON = "scrobbles.json" "

Yeah, I think I should delete that from the script later. Just keep it in, but when you want to re-run it later, remove that file so it recreates it.

**Reply by zeroquinc** [reply]
I guess I could also make more configuration options like how much it seeks back, now for every run it syncs your whole last.fm history, but when you want to run it once every week, it makes more sense to set up a file with a last import date and use that so that people with a lot of scrobbles don't have to hit the last.fm API that much. I will take another look at it soon.

I even made a new script that added the play count of my other last.fm account. My Navidrome db now has the most accurate play count out of them all 😂

```
Incremented Breaking Benjamin - breath by 67 plays (now 78)
Incremented Coldplay - christmas lights by 109 plays (now 136)
Incremented Intervals - meridian by 87 plays (now 102)
Incremented Suzan & Freek - weg van jou by 52 plays (now 65)
Incremented Snelle - mini cooper by 26 plays (now 41)
Incremented OneRepublic - secrets by 28 plays (now 48)
Incremented 2Pac - life goes on by 69 plays (now 97)
```

**Reply by KruyKaze** [reply]
> I guess I could also make more configuration options like how much it seeks back, now for every run it syncs your whole last.fm history, but when you want to run it once every week, it makes more sense to set up a file with a last import date and use that so that people with a lot of scrobbles don't have to hit the last.fm API that much. I will take another look at it soon.

Actually that wouldn't work when you're rebuiling your library. Think of the cases where you add a track later to your Navidrome and your script doesn't check old scrobbles anymore. Does that make sense?

**Reply by zeroquinc** [reply]
Yeah makes sense. I will take another look at it later. For now just remove the scrobble json every time it runs. It will never replace it with last.fm if Navidrome db has more play count. Which in theory could now never happen for me since I only use Navidrome for listening and scrobble directly to last.fm.

Did the script work for you?

**Reply by KruyKaze** [reply]
> Yeah makes sense. I will take another look at it later. For now just remove the scrobble json every time it runs. It will never replace it with last.fm if Navidrome db has more play count. Which in theory could now never happen for me since I only use Navidrome for listening and scrobble directly to last.fm.
> 
> Did the script work for you?

It did ! You solved a huge problem for me thank you!

**Reply by zeroquinc** [reply]
Added another update that normalizes the artist name, before it only did the track name. It matched even more scrobbles for me.

For example I noticed "French the Kid" didn't match "French The Kid". Now it does.

**Reply by KruyKaze** [reply]
> Added another update that normalizes the artist name, before it only did the track name. It matched even more scrobbles for me.
> 
> For example I noticed "French the Kid" didn't match "French The Kid". Now it does.

Maybe i'm asking too much but is it possible for your script to tell me which albums/artists i'm missing from my favorites in lastfm?

**Reply by zeroquinc** [reply]
> > Added another update that normalizes the artist name, before it only did the track name. It matched even more scrobbles for me.
> > For example I noticed "French the Kid" didn't match "French The Kid". Now it does.
> 
> Maybe i'm asking too much but is it possible for your script to tell me which albums/artists i'm missing from my favorites in lastfm?

Hehe no you are not asking to much, that's a great idea! 
Good way to start rebuilding the library again.

I have added it to the script. It will be saved as: missing_from_scrobbles.json and missing_from_loved.json. Make sure to delete scrobbles.json when running it again. Still need to delete that.

Only downside is that the json will show it all in lowercase and it will only fetch your loved tracks if you scrobbled the track at least once (but why is it even loved if you have never scrobbled it before haha). It's also not possible (yet) to show album names, because the whole script is focused around artist + track. Maybe later I will modify this all and put it in an actual repo. For now it does the job.

missing_from_loved.json example.
```
{
  "hef": [
    {
      "track": "fontein"
    }
  ]
}
```

**Reply by zeroquinc** [reply]
> > Added another update that normalizes the artist name, before it only did the track name. It matched even more scrobbles for me.
> > For example I noticed "French the Kid" didn't match "French The Kid". Now it does.
> 
> Maybe i'm asking too much but is it possible for your script to tell me which albums/artists i'm missing from my favorites in lastfm?

Because it's actually pretty useful to have the album name too, I edited the script to include the album. It stores it in the json like this:

```
{
  "howard shore": {
    "The Lord of the Rings: The Return of the King: Original Motion Picture Soundtrack": [
      {
        "track": "andúril",
        "loved": false
      }
    ],
    "The Lord of the Rings: The Two Towers: The Complete Recordings": [
      {
        "track": "one of the dúnedain / evenstar",
        "loved": false
      },
      {
        "track": "ent‐draught",
        "loved": false
      },
 ```

**Reply by KruyKaze** [reply]
> > > Added another update that normalizes the artist name, before it only did the track name. It matched even more scrobbles for me.
> > > For example I noticed "French the Kid" didn't match "French The Kid". Now it does.
> > 
> > 
> > Maybe i'm asking too much but is it possible for your script to tell me which albums/artists i'm missing from my favorites in lastfm?
> 
> Because it's actually pretty useful to have the album name too, I edited the script to include the album. It stores it in the json like this:


Very nice , let me give it another run now.

**Reply by zeroquinc** [reply]
> > > > Added another update that normalizes the artist name, before it only did the track name. It matched even more scrobbles for me.
> > > > For example I noticed "French the Kid" didn't match "French The Kid". Now it does.
> > > 
> > > 
> > > Maybe i'm asking too much but is it possible for your script to tell me which albums/artists i'm missing from my favorites in lastfm?
> > 
> > 
> > Because it's actually pretty useful to have the album name too, I edited the script to include the album. It stores it in the json like this:
> 
> Very nice , let me give it another run now.

Wait one sec, I'm running it now too, I've also actually managed to make the artist and track the original value too, with capitalization. So it looks better in the json. Will upload it any sec.

EDIT: Updated, now jsons looks like this, this is the last update for now. Curious to hear how it worked out for you.

```
"UB40": {
    "UB40 Collected": [
      {
        "track": "I Got You Babe - Remastered",
        "loved": false
      },
      {
        "track": "Kingston Town - Remastered",
        "loved": false
      },
      {
        "track": "Homely Girl - Remastered",
        "loved": false
      }
    ],
    "Labour of Love II": [
      {
        "track": "Wear You To The Ball",
        "loved": false
      }
    ],
    "Labour Of Love": [
      {
        "track": "Keep On Moving",
        "loved": false
      },
      {
        "track": "Please Don’t Make Me Cry",
        "loved": false
      },
```

**Reply by KruyKaze** [reply]
> > > > > Added another update that normalizes the artist name, before it only did the track name. It matched even more scrobbles for me.
> > > > > For example I noticed "French the Kid" didn't match "French The Kid". Now it does.
> > > > 
> > > > 
> > > > Maybe i'm asking too much but is it possible for your script to tell me which albums/artists i'm missing from my favorites in lastfm?
> > > 
> > > 
> > > Because it's actually pretty useful to have the album name too, I edited the script to include the album. It stores it in the json like this:
> > 
> > 
> > Very nice , let me give it another run now.
> 
> Wait one sec, I'm running it now too, I've also actually managed to make the artist and track the original value too, with capitalization. So it looks better in the json. Will upload it any sec.
> 

Too late lol

> EDIT: Updated, now jsons looks like this, this is the last update for now. Curious to hear how it worked out for you.
> 
> ```
> "UB40": {
>     "UB40 Collected": [
>       {
>         "track": "I Got You Babe - Remastered",
>         "loved": false
>       },
>       {
>         "track": "Kingston Town - Remastered",
>         "loved": false
>       },
>       {
>         "track": "Homely Girl - Remastered",
>         "loved": false
>       }
>     ],
>     "Labour of Love II": [
>       {
>         "track": "Wear You To The Ball",
>         "loved": false
>       }
>     ],
>     "Labour Of Love": [
>       {
>         "track": "Keep On Moving",
>         "loved": false
>       },
>       {
>         "track": "Please Don’t Make Me Cry",
>         "loved": false
>       },
> ```

even better running again now

**Reply by zeroquinc** [reply]
> even better running again now

Made one small fix again, that somehow got removed in a revision, that shows what will actually change before you press Y in the script.

```
  - UB40 - Red Red Wine
    Last.fm: 14 | Navidrome: 0 | Diff: +14 | Loved: False

  - UB40 - (I Can't Help) Falling In Love With You
    Last.fm: 28 | Navidrome: 0 | Diff: +28 | Loved: False
    
    Proceed with updating these playcounts and stars in Navidrome? [y/N]:
```

**Reply by zeroquinc** [reply]
One more update lol. 

- now the jsons also show the amount of times the track has been scrobbled and the date it has been last played on last.fm. (I should make a repo lol)

- If the plays in the Navidrome db is more than on last.fm, it will ask to overwrite it with last.fm scrobbles.

- if the date of last played in the Navidrome db is newer than last.fm, it will always use the latest played date.

```
{
  "Howard Shore": {
    "The Lord of the Rings: The Return of the King: Original Motion Picture Soundtrack": [
      {
        "track": "Andúril",
        "scrobbled": 1,
        "loved": false,
        "lastplayed": "2025-10-20 20:58:45"
      }
    ],
```

**Reply by KruyKaze** [reply]
> One more update lol.
> 
>     * now the jsons also show the amount of times the track has been scrobbled and the date it has been last played on last.fm. (I should make a repo lol)
> 
The repo is a good idea I think

**Reply by zeroquinc** [reply]
I made a repo 👍 

https://github.com/zeroquinc/NaviSync

**Reply by CustardFlan** [reply]
Works very well, thank you for the code!

**Reply by zeroquinc** [reply]
> Works very well, thank you for the code!

No problem, be sure to star or follow the repo, I'm currently adding in API support, meaning you don't need to stop Navidrome to read the database.

**Reply by dewey** [reply]
Hey @zeroquinc, what did you mean with "adding API support"? From what I could see there's no Navidrome API to set the play count directly to a value, only to read it.

I was think of opening a discussion to see if that would be something that would be considered as an addition to the existing API. What do you think?

**Reply by zeroquinc** [reply]
> Hey @zeroquinc, what did you mean with "adding API support"? From what I could see there's no Navidrome API to set the play count directly to a value, only to read it.
> 
> I was think of opening a discussion to see if that would be something that would be considered as an addition to the existing API. What do you think?

Hey!

My script used to have a API mode, it send a lot of "scrobbles" to Navidrome. 
But it was really buggy as you can imagine, because if you played a song 120 times, I had to "fake" a scrobble and that 120 times. It made the database go a bit freaky. If we could update the play count directly with a value without sending multiple fake plays, it would be a nice addition to add to the script.

Also then it could be maybe made into a plugin for Navidrome, instead of a standalone script. I cant modify it into a plugin now cause Navidrome has to be closed for it to work.

So yeah by all means open a discussion for it!

**Reply by dewey** [reply]
Ah nice, will do then! That's exactly the API I also found but thought it would be a bit too hacky 😬

**Reply by zeroquinc** [reply]
> Ah nice, will do then! That's exactly the API I also found but thought it would be a bit too hacky 😬

Yeah it was really buggy. Also if you had a Last.FM or any other scrobbler enabled it would forward the fake plays to that, making it a real mess. I since then removed it and only support the database mode, since we modify the database directly there.

If you could make a feature request for it, I will be sure to upvote it.

**Reply by dewey** [reply]
I've started a top level discussion about that proposal: https://github.com/navidrome/navidrome/discussions/5282

**Comment by schuerik** [comment]
Would be nice to have that for all the scrobble services that are already supported by navidrome. I think the same could be done for listenbrainz. 
For once that would serve as a great backup system, in case of issues with the navidrome database (which happened way to often to me now and each time all my statistics get reset) and also I've imported all of my spotify history (>250k scrobbles) into listenbrainz. I could get way better recommendations and top song lists, if navidrome would just know all of my listening history.
